File: game-engine/src/main/kotlin/com/failureludo/engine/GameEngine.kt
package com.failureludo.engine

/**
 * The central coordinator that drives game flow.
 *
 * All public functions are pure (return a new [GameState]) so they can be called
 * from any thread, stored, and later replayed or transmitted over a network.
 *
 * Typical call sequence:
 *   1. [rollDice]  — on each turn, generates a dice value and evaluates movable pieces
 *   2. [selectPiece] — player chooses which piece to move
 *   3. Repeat until [GameState.isGameOver]
 */
object GameEngine {

    private val dice = DiceRoller()

    // ── Factory ──────────────────────────────────────────────────────────────

    /**
     * Creates a fresh [GameState] for a new game.
     *
     * @param activeColors   Which of the 4 colours are active (2-4 items).
     *                       Order determines turn order.
     * @param playerTypes    Map of colour → human or bot.
     * @param playerNames    Custom display names (optional).
     * @param mode           FFA or Team.
     */
    fun newGame(
        activeColors: List<PlayerColor>,
        playerTypes: Map<PlayerColor, PlayerType> = emptyMap(),
        playerNames: Map<PlayerColor, String> = emptyMap(),
        mode: GameMode = GameMode.FREE_FOR_ALL
    ): GameState {
        require(activeColors.size in 2..4) { "Need between 2 and 4 active colours." }
        if (mode == GameMode.TEAM) {
            require(activeColors.size == 4) { "Team mode requires exactly 4 players." }
        }

        // All 4 colours are always created; inactive ones are just marked inactive
        val players = PlayerColor.entries.map { color ->
            Player(
                id = PlayerId(color.ordinal + 1),
                color = color,
                name = playerNames[color] ?: "Player-${color.ordinal + 1}",
                type = playerTypes[color] ?: PlayerType.HUMAN,
                isActive = color in activeColors
            )
        }

        // First turn: first ACTIVE player in the activeColors list
        val startIndex = players.indexOfFirst { it.color == activeColors.first() }

        return GameState(
            players = players,
            mode = mode,
            moveCounter = 0L,
            currentPlayerIndex = startIndex,
            turnPhase = TurnPhase.WAITING_FOR_ROLL,
            diceByPlayer = PlayerColor.entries.associate { PlayerId(it.ordinal + 1) to null }
        )
    }

    // ── Turn Actions ─────────────────────────────────────────────────────────

    /**
     * Applies a full deterministic turn for replay/import use-cases.
     *
     * This helper enforces actor ownership, injects a known dice value, and then
     * applies the chosen piece movement in one operation.
     */
    fun applyDeterministicTurn(state: GameState, input: DeterministicTurnInput): GameState {
        require(state.turnPhase == TurnPhase.WAITING_FOR_ROLL) {
            "applyDeterministicTurn called in wrong phase: ${state.turnPhase}"
        }
        require(state.currentPlayer.id == input.actorId) {
            "Actor ${input.actorId.value} is not the current player ${state.currentPlayer.id.value}."
        }

        val rolled = rollDice(state, input.diceValue)
        require(rolled.turnPhase == TurnPhase.WAITING_FOR_PIECE_SELECTION) {
            "Deterministic move requires a selectable piece, but phase is ${rolled.turnPhase}."
        }

        val movingPlayer = requireNotNull(rolled.players.firstOrNull { it.id == input.movingPlayerId }) {
            "Moving player ${input.movingPlayerId.value} not found."
        }
        val selectedPiece = requireNotNull(
            rolled.movablePieces.firstOrNull { piece ->
                piece.color == movingPlayer.color && piece.id == input.pieceId
            }
        ) {
            "Piece ${input.pieceId} for player ${input.movingPlayerId.value} is not movable."
        }

        return selectPiece(rolled, selectedPiece, input.deferHomeEntry)
    }

    /**
     * Applies a deterministic roll that should not produce a piece selection.
     *
     * Used for replay entries where a player rolled but had no legal move, or
     * forfeited via three consecutive sixes.
     */
    fun applyDeterministicRollOnly(state: GameState, actorId: PlayerId, diceValue: Int): GameState {
        require(state.turnPhase == TurnPhase.WAITING_FOR_ROLL) {
            "applyDeterministicRollOnly called in wrong phase: ${state.turnPhase}"
        }
        require(state.currentPlayer.id == actorId) {
            "Actor ${actorId.value} is not the current player ${state.currentPlayer.id.value}."
        }

        val rolled = rollDice(state, diceValue)
        return when (rolled.turnPhase) {
            TurnPhase.NO_MOVES_AVAILABLE -> advanceNoMoves(rolled)
            TurnPhase.WAITING_FOR_ROLL -> rolled // consecutive-sixes forfeit already advanced turn
            TurnPhase.WAITING_FOR_PIECE_SELECTION -> {
                throw IllegalArgumentException(
                    "Deterministic roll-only is invalid because movable pieces exist."
                )
            }
            TurnPhase.GAME_OVER -> rolled
        }
    }

    /**
     * Rolls the dice for the current player.
     * Updates [TurnPhase] to WAITING_FOR_PIECE_SELECTION or NO_MOVES_AVAILABLE.
     */
    fun rollDice(state: GameState): GameState {
        require(state.turnPhase == TurnPhase.WAITING_FOR_ROLL) {
            "rollDice called in wrong phase: ${state.turnPhase}"
        }

        return rollDiceWithValue(state, dice.roll())
    }

    /**
     * Rolls with a forced dice value for deterministic replay/import paths.
     */
    fun rollDice(state: GameState, forcedDiceValue: Int): GameState {
        require(state.turnPhase == TurnPhase.WAITING_FOR_ROLL) {
            "rollDice called in wrong phase: ${state.turnPhase}"
        }
        require(forcedDiceValue in 1..6) {
            "Forced dice value must be 1..6, got $forcedDiceValue."
        }

        return rollDiceWithValue(state, forcedDiceValue)
    }

    private fun rollDiceWithValue(state: GameState, diceValue: Int): GameState {
        val rollCount = (state.lastDice?.rollCount?.takeIf { diceValue == 6 && it < 3 } ?: 0) + 1

        // Three consecutive 6s → forfeit turn
        if (diceValue == 6 && rollCount == 3) {
            val event = GameEvent.ConsecutiveSixesForfeit(
                playerId = state.currentPlayer.id,
                color = state.currentPlayer.color
            )
            val updatedDiceByPlayer = state.diceByPlayer.toMutableMap().apply {
                this[state.currentPlayer.id] = diceValue
            }
            return advanceToNextTurn(
                state.copy(
                    lastDice = DiceResult(diceValue, rollCount),
                    diceByPlayer = updatedDiceByPlayer,
                    eventLog = state.eventLog + event
                )
            )
        }

        val diceResult = DiceResult(diceValue, rollCount)
        val movable = GameRules.movablePiecesForTurn(
            currentPlayer = state.currentPlayer,
            diceValue = diceValue,
            allPlayers = state.players,
            mode = state.mode,
            sharedTeamDiceEnabled = state.sharedTeamDiceEnabled
        )

        val newPhase = if (movable.isEmpty()) TurnPhase.NO_MOVES_AVAILABLE
                       else TurnPhase.WAITING_FOR_PIECE_SELECTION

        val updatedDiceByPlayer = state.diceByPlayer.toMutableMap().apply {
            this[state.currentPlayer.id] = diceValue
        }

        return state.copy(
            lastDice = diceResult,
            diceByPlayer = updatedDiceByPlayer,
            movablePieces = movable,
            turnPhase = newPhase
        )
    }

    /**
     * Auto-advances when there are no moves (caller should invoke after a short UI delay).
     */
    fun advanceNoMoves(state: GameState): GameState {
        require(state.turnPhase == TurnPhase.NO_MOVES_AVAILABLE)
        val event = GameEvent.TurnSkipped(
            playerId = state.currentPlayer.id,
            color = state.currentPlayer.color
        )
        return advanceToNextTurn(state.copy(eventLog = state.eventLog + event))
    }

    /**
     * Moves the selected [piece] using the current dice value.
     *
     * Returns the updated [GameState], which may include:
     * - A capture event
     * - An extra-roll turn (dice == 6 or capture occurred)
     * - A win condition
     * - Normal turn advancement
     */
    fun selectPiece(state: GameState, piece: Piece, deferHomeEntry: Boolean = false): GameState {
        require(state.turnPhase == TurnPhase.WAITING_FOR_PIECE_SELECTION) {
            "selectPiece called in wrong phase: ${state.turnPhase}"
        }
        require(piece in state.movablePieces) { "Piece ${piece.id} is not movable." }

        val diceValue = state.lastDice!!.value
        val movingPlayer = state.players.firstOrNull { it.color == piece.color }
            ?: error("Moving player not found for color ${piece.color}")
        val movingPlayerId = movingPlayer.id
        val movingColor = piece.color
        val actingPlayerId = state.currentPlayer.id
        val actingColor = state.currentPlayer.color

        // Compute destination
        val destination = GameRules.computeDestination(
            piece,
            diceValue,
            movingColor,
            state.players,
            state.mode,
            deferHomeEntry
        )!!

        // Count captures BEFORE applying move
        val captureTargets = GameRules.captureTargets(
            piece,
            diceValue,
            movingColor,
            state.players,
            state.mode,
            deferHomeEntry
        )
        val captureCount = captureTargets.size

        // Apply the move
        val newMoveCounter = state.moveCounter + 1
        val newPlayers = GameRules.applyMove(
            piece = piece,
            diceValue = diceValue,
            color = movingColor,
            players = state.players,
            mode = state.mode,
            deferHomeEntry = deferHomeEntry,
            movedAt = newMoveCounter
        )

        // Build event list
        val events = mutableListOf<GameEvent>()
        val isEntering = piece.position is PiecePosition.HomeBase
        if (isEntering) {
            events += GameEvent.PieceEnteredBoard(
                playerId = movingPlayerId,
                color = movingColor,
                pieceId = piece.id
            )
        } else {
            events += GameEvent.PieceMoved(playerId = movingPlayerId, color = movingColor, pieceId = piece.id)
        }
        if (destination is PiecePosition.Finished) {
            events += GameEvent.PieceFinished(
                playerId = movingPlayerId,
                color = movingColor,
                pieceId = piece.id
            )
        }

        // Capture events
        captureTargets.forEach { target ->
            val capturedPlayerId = state.players.firstOrNull { it.color == target.color }?.id
                ?: error("Captured player not found for color ${target.color}")
            events += GameEvent.PieceCaptured(
                capturedPlayerId = capturedPlayerId,
                capturedColor = target.color,
                byPlayerId = actingPlayerId,
                byColor = movingColor
            )
        }

        val updatedEnteredBoardFlags = state.hasEnteredBoardAtLeastOnce.toMutableMap().apply {
            if (isEntering) {
                this[movingPlayerId] = true
            }
        }

        val newState = state.copy(
            players = newPlayers,
            moveCounter = newMoveCounter,
            eventLog = state.eventLog + events,
            hasEnteredBoardAtLeastOnce = updatedEnteredBoardFlags,
            movablePieces = emptyList()
        )

        // Check winner
        val winners = GameRules.checkWinner(newPlayers, state.mode)
        if (winners != null) {
            return newState.copy(
                winners = winners,
                turnPhase = TurnPhase.GAME_OVER,
                eventLog = newState.eventLog + GameEvent.PlayerWon(winners)
            )
        }

        // Extra roll?
        val captured = captureCount > 0
        val extraRoll = GameRules.grantsExtraRoll(diceValue, piece, newPlayers[state.currentPlayerIndex].pieces, captured)
        return if (extraRoll) {
            val reason = if (diceValue == 6) "rolled a 6" else "captured a piece"
            newState.copy(
                turnPhase = TurnPhase.WAITING_FOR_ROLL,
                lastDice = DiceResult(diceValue, state.lastDice.rollCount),
                eventLog = newState.eventLog + GameEvent.ExtraRollGranted(
                    playerId = actingPlayerId,
                    color = actingColor,
                    reason = reason
                )
            )
        } else {
            advanceToNextTurn(newState)
        }
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun advanceToNextTurn(state: GameState): GameState {
        val nextIdx = GameRules.nextPlayerIndex(state.currentPlayerIndex, state.players)
        val sharedTeamDiceEnabled = computeSharedTeamDiceEnabled(state)
        return state.copy(
            currentPlayerIndex = nextIdx,
            turnPhase = TurnPhase.WAITING_FOR_ROLL,
            lastDice = null,
            sharedTeamDiceEnabled = sharedTeamDiceEnabled
        )
    }

    private fun computeSharedTeamDiceEnabled(state: GameState): Set<Int> {
        if (state.mode != GameMode.TEAM) return emptySet()

        val unlocked = state.sharedTeamDiceEnabled.toMutableSet()
        val activeByTeam = state.players
            .filter { it.isActive }
            .groupBy { it.color.teamIndex }

        activeByTeam.forEach { (teamIndex, teamPlayers) ->
            if (teamPlayers.size < 2) return@forEach
            val bothHaveEntered = teamPlayers.all { teammate ->
                state.hasEnteredBoardAtLeastOnce[teammate.id] == true
            }
            if (bothHaveEntered) {
                unlocked += teamIndex
            }
        }

        return unlocked
    }

}

File: ludo-web/src/engine/gameEngine.ts
import {
  type DeterministicTurnInput,
  type DiceResult,
  type GameEvent,
  type GameMode,
  type GameState,
  type Piece,
  type Player,
  type PlayerColor,
  HomeBase,
  currentPlayer,
  ALL_COLORS,
} from './types';
import { colorIndex, teamIndex } from './board';
import {
  applyMove,
  captureTargets,
  checkWinner,
  computeDestination,
  grantsExtraRoll,
  movablePiecesForTurn,
  nextPlayerIndex,
} from './gameRules';

function makePiece(id: number, color: PlayerColor): Piece {
  return { id, color, position: HomeBase, lastMovedAt: 0 };
}

function makePlayer(color: PlayerColor, name: string, isActive: boolean): Player {
  return {
    id: colorIndex(color) + 1,
    color,
    name,
    pieces: [0, 1, 2, 3].map(id => makePiece(id, color)),
    isActive,
  };
}

function initialDiceByPlayer(): Record<number, number | null> {
  const r: Record<number, number | null> = {};
  for (const c of ALL_COLORS) r[colorIndex(c) + 1] = null;
  return r;
}

function initialEnteredFlags(): Record<number, boolean> {
  const r: Record<number, boolean> = {};
  for (const c of ALL_COLORS) r[colorIndex(c) + 1] = false;
  return r;
}

export function newGame(
  activeColors: PlayerColor[],
  playerNames: Partial<Record<PlayerColor, string>> = {},
  mode: GameMode = 'FREE_FOR_ALL',
): GameState {
  const players = ALL_COLORS.map(c =>
    makePlayer(c, playerNames[c] ?? `Player-${colorIndex(c) + 1}`, activeColors.includes(c)),
  );
  const startIndex = players.findIndex(p => p.color === activeColors[0]);
  return {
    players,
    mode,
    moveCounter: 0,
    currentPlayerIndex: startIndex,
    turnPhase: 'WAITING_FOR_ROLL',
    lastDice: null,
    diceByPlayer: initialDiceByPlayer(),
    hasEnteredBoardAtLeastOnce: initialEnteredFlags(),
    sharedTeamDiceEnabled: new Set(),
    movablePieces: [],
    winners: null,
    eventLog: [],
  };
}

function rollDiceWithValue(state: GameState, diceValue: number): GameState {
  const player = currentPlayer(state);
  const prevRollCount = state.lastDice?.rollCount ?? 0;
  const newRollCount = diceValue === 6 ? prevRollCount + 1 : 1;

  if (diceValue === 6 && newRollCount === 3) {
    const event: GameEvent = { type: 'ConsecutiveSixesForfeit', playerId: player.id, color: player.color };
    const updated = {
      ...state,
      lastDice: { value: diceValue, rollCount: newRollCount } as DiceResult,
      diceByPlayer: { ...state.diceByPlayer, [player.id]: diceValue },
      eventLog: [...state.eventLog, event],
    };
    return advanceToNextTurn(updated);
  }

  const movable = movablePiecesForTurn(player, diceValue, state.players, state.mode, state.sharedTeamDiceEnabled);
  const newPhase = movable.length === 0 ? 'NO_MOVES_AVAILABLE' : 'WAITING_FOR_PIECE_SELECTION';

  return {
    ...state,
    lastDice: { value: diceValue, rollCount: newRollCount },
    diceByPlayer: { ...state.diceByPlayer, [player.id]: diceValue },
    movablePieces: movable,
    turnPhase: newPhase,
  };
}

export function rollDice(state: GameState, forcedValue: number): GameState {
  return rollDiceWithValue(state, forcedValue);
}

export function advanceNoMoves(state: GameState): GameState {
  const player = currentPlayer(state);
  const event: GameEvent = { type: 'TurnSkipped', playerId: player.id, color: player.color };
  return advanceToNextTurn({ ...state, eventLog: [...state.eventLog, event] });
}

export function selectPiece(state: GameState, piece: Piece, deferHomeEntry = false): GameState {
  const diceValue = state.lastDice!.value;
  const movingPlayer = state.players.find(p => p.color === piece.color)!;
  const actingPlayer = currentPlayer(state);

  const captures = captureTargets(piece, diceValue, piece.color, state.players, state.mode, deferHomeEntry);
  const captureCount = captures.length;
  const newMoveCounter = state.moveCounter + 1;

  const newPlayers = applyMove(
    piece, diceValue, piece.color, state.players, state.mode, deferHomeEntry, newMoveCounter,
  );

  const events: GameEvent[] = [];
  const isEntering = piece.position.type === 'HomeBase';
  const destination = computeDestination(piece, diceValue, piece.color, state.players, state.mode, deferHomeEntry)!;

  if (isEntering) {
    events.push({ type: 'PieceEnteredBoard', playerId: movingPlayer.id, color: piece.color, pieceId: piece.id });
  } else {
    events.push({ type: 'PieceMoved', playerId: movingPlayer.id, color: piece.color, pieceId: piece.id });
  }
  if (destination.type === 'Finished') {
    events.push({ type: 'PieceFinished', playerId: movingPlayer.id, color: piece.color, pieceId: piece.id });
  }
  for (const c of captures) {
    const capturedPlayer = state.players.find(p => p.color === c.color)!;
    events.push({
      type: 'PieceCaptured',
      capturedPlayerId: capturedPlayer.id,
      capturedColor: c.color,
      byPlayerId: actingPlayer.id,
      byColor: piece.color,
    });
  }

  const updatedEntered = {
    ...state.hasEnteredBoardAtLeastOnce,
    ...(isEntering ? { [movingPlayer.id]: true } : {}),
  };

  const newState: GameState = {
    ...state,
    players: newPlayers,
    moveCounter: newMoveCounter,
    eventLog: [...state.eventLog, ...events],
    hasEnteredBoardAtLeastOnce: updatedEntered,
    movablePieces: [],
  };

  const winners = checkWinner(newPlayers, state.mode);
  if (winners) {
    return {
      ...newState,
      winners,
      turnPhase: 'GAME_OVER',
      eventLog: [...newState.eventLog, { type: 'PlayerWon', playerIds: winners }],
    };
  }

  const extra = grantsExtraRoll(diceValue, captureCount > 0);
  if (extra) {
    const reason = diceValue === 6 ? 'rolled a 6' : 'captured a piece';
    return {
      ...newState,
      turnPhase: 'WAITING_FOR_ROLL',
      lastDice: state.lastDice,
      eventLog: [
        ...newState.eventLog,
        { type: 'ExtraRollGranted', playerId: actingPlayer.id, color: actingPlayer.color, reason },
      ],
    };
  }

  return advanceToNextTurn(newState);
}

export function applyDeterministicTurn(state: GameState, input: DeterministicTurnInput): GameState {
  const rolled = rollDice(state, input.diceValue);
  const movingPlayer = rolled.players.find(p => p.id === input.movingPlayerId)!;
  const selectedPiece = rolled.movablePieces.find(
    p => p.color === movingPlayer.color && p.id === input.pieceId,
  );
  if (!selectedPiece) throw new Error(`Piece ${input.pieceId} for player ${input.movingPlayerId} not movable`);
  return selectPiece(rolled, selectedPiece, input.deferHomeEntry);
}

export function applyDeterministicRollOnly(state: GameState, _actorId: number, diceValue: number): GameState {
  const rolled = rollDice(state, diceValue);
  if (rolled.turnPhase === 'NO_MOVES_AVAILABLE') return advanceNoMoves(rolled);
  if (rolled.turnPhase === 'WAITING_FOR_ROLL') return rolled; // consecutive-sixes forfeit
  throw new Error(`applyDeterministicRollOnly: unexpected phase ${rolled.turnPhase}`);
}

function advanceToNextTurn(state: GameState): GameState {
  const nextIdx = nextPlayerIndex(state.currentPlayerIndex, state.players);
  return {
    ...state,
    currentPlayerIndex: nextIdx,
    turnPhase: 'WAITING_FOR_ROLL',
    lastDice: null,
    sharedTeamDiceEnabled: computeSharedTeamDiceEnabled(state),
  };
}

function computeSharedTeamDiceEnabled(state: GameState): Set<number> {
  if (state.mode !== 'TEAM') return new Set();
  const unlocked = new Set(state.sharedTeamDiceEnabled);
  const activeByTeam = new Map<number, Player[]>();
  for (const p of state.players.filter(p => p.isActive)) {
    const ti = teamIndex(p.color);
    if (!activeByTeam.has(ti)) activeByTeam.set(ti, []);
    activeByTeam.get(ti)!.push(p);
  }
  for (const [ti, team] of activeByTeam) {
    if (team.length < 2) continue;
    if (team.every(p => state.hasEnteredBoardAtLeastOnce[p.id])) unlocked.add(ti);
  }
  return unlocked;
}
